### PR TITLE
[feat] Add additional CFI checks to firmware image library

### LINKED ISSUE #982: Fault injection hardening for verifier code
3 asks:
1. Ensure no constant time window between externally-visible signal and critical branch.
2. Add/confirm init-to-fail logic for boolean checks
3. Use multi-bit booleans

Details are captured in https://github.com/chipsalliance/caliptra-sw/blob/main/rom/dev/doc/threat-model.md#mitigation-2

### PR DESCRIPTION
**Pull request by mhatrevi**
This change addresses item 1 from the issue# https://github.com/chipsalliance/caliptra-sw/issues/982

### CODE REVIEW

File: image/verify/src/verifier.rs [hunk 1]
```diff
@@ -216,24 +221,32 @@ impl<Env: ImageVerificationEnv> ImageVerifier<Env> {
 
         match key_idx {
             0..=SECOND_LAST_KEY_IDX => {
+                cfi_assert_le(cfi_launder(key_idx), SECOND_LAST_KEY_IDX);
                 let key = VendorPubKeyRevocation::from_bits_truncate(0x01u32 << key_idx);
-                if revocation.contains(key) {
+                if cfi_launder(revocation.contains(key)) {
                     Err(CaliptraError::IMAGE_VERIFIER_ERR_VENDOR_ECC_PUB_KEY_REVOKED)?;
+                } else {
+                    cfi_assert!(!revocation.contains(key));
```
Comment (korran) [inline image/verify/src/verifier.rs:229]: Need to double check the assembly that this re-does the full contains operations instead of just recomparing the result (which is what I suspect).

Comment (mhatrevi) [reply image/verify/src/verifier.rs:229]: Yeah, revocation should be laundered. Hardened it even further by laundering the key as well.

image

File: fmc/tests/fmc_integration_tests/test_hand_off.rs [hunk 2]
```diff
@@ -4,7 +4,7 @@ use caliptra_hw_model::{BootParams, HwModel, InitParams};
 
 #[test]
 fn test_hand_off() {
-    let rom = caliptra_builder::build_firmware_rom(&firmware::ROM_WITH_UART).unwrap();
+    let rom = caliptra_builder::build_firmware_rom(&firmware::ROM_WITH_UART_NO_CFI).unwrap();
```
Comment (korran) [inline fmc/tests/fmc_integration_tests/test_hand_off.rs:7]: revert this change

File: test/tests/caliptra_integration_tests/smoke_test.rs [hunk 3]
```diff
@@ -541,7 +541,7 @@ fn fips_cmd_test_rt() {
         .set_debug_locked(false)
         .set_device_lifecycle(DeviceLifecycle::Unprovisioned);
 
-    let rom = caliptra_builder::build_firmware_rom(&firmware::ROM_WITH_UART).unwrap();
+    let rom = caliptra_builder::build_firmware_rom(&firmware::ROM_WITH_UART_NO_CFI).unwrap();
```
Comment (korran) [inline test/tests/caliptra_integration_tests/smoke_test.rs:544]: revert this?

File: builder/src/firmware.rs [hunk 4]
```diff
@@ -17,6 +17,12 @@ pub const ROM_WITH_UART: FwId = FwId {
     features: &["emu"],
 };
 
+pub const ROM_WITH_UART_NO_CFI: FwId = FwId {
```
Comment (korran) [inline builder/src/firmware.rs:20]: Either remove this or add it to REGISTERED_FW

File: image/verify/src/verifier.rs [hunk 5]
```diff
@@ -216,24 +226,32 @@ impl<Env: ImageVerificationEnv> ImageVerifier<Env> {
...
-        if reason == ResetReason::UpdateReset {
+        if cfi_launder(reason) == ResetReason::UpdateReset {
             let expected = self.env.vendor_ecc_pub_key_idx_dv();
-            if expected != key_idx {
+            if cfi_launder(expected) != key_idx {
                 Err(
                     CaliptraError::IMAGE_VERIFIER_ERR_UPDATE_RESET_VENDOR_ECC_PUB_KEY_IDX_MISMATCH,
                 )?;
+            } else {
+                cfi_assert_eq(expected, key_idx);
```
Comment (korran) [inline image/verify/src/verifier.rs:251]: We should be going back to the source here:

```
cfi_assert_eq(self.env.vendor_ecc_pub_key_idx_dv(), key_idx);
```

File: image/verify/src/verifier.rs [hunk 6]
```diff
@@ -253,23 +271,31 @@ impl<Env: ImageVerificationEnv> ImageVerifier<Env> {
...
-        if reason == ResetReason::UpdateReset {
+        if cfi_launder(reason) == ResetReason::UpdateReset {
             let expected = self.env.vendor_lms_pub_key_idx_dv();
-            if expected != key_idx {
+            if cfi_launder(expected) != key_idx {
                 Err(
                     CaliptraError::IMAGE_VERIFIER_ERR_UPDATE_RESET_VENDOR_LMS_PUB_KEY_IDX_MISMATCH,
                 )?;
+            } else {
+                cfi_assert_eq(expected, key_idx);
```
Comment (korran) [inline image/verify/src/verifier.rs:295]: Probably worth going back to the source here:

```
cfi_assert_eq(self.env.vendor_lms_pub_key_idx_dv(), key_idx);
```

File: image/verify/src/verifier.rs [hunk 7]
```diff
@@ -760,15 +835,18 @@ impl<Env: ImageVerificationEnv> ImageVerifier<Env> {
...
             0_u32
         } else {
+            cfi_assert!(!anti_rollback_disable);
             fuse_svn
         }
     }
 }
 
 #[cfg(all(test, target_family = "unix"))]
+#[cfg(feature = "no-cfi")]
```
Comment (korran) [inline image/verify/src/verifier.rs:849]: This is dangerous. With this change, `cargo test` no longer runs these tests.

Comment (korran) [reply image/verify/src/verifier.rs:849]: I'm not sure we have a lot of options to fix this. The first one that comes to mind is to define CFI_STATE_ORG and cfi_panic_handler

**Review by korran — APPROVED** [review]

**Review by jhand2 — APPROVED** [review]
